mmcv/ops/csrc/common/cuda/softmax_focal_loss_cuda_kernel.cuh

```diff
@@ -10,62 +10,125 @@
+    // dFL/dp = - [((1-p)^gamma) / p
+    //             - gamma * (1-p)^(gamma-1) * log(p)]
+    // dp_i/dx_j = dSM/dx_j
+    //           = p_i * (1-p_j) i==j;
+    //             p_i * (0-p_j) i!=j;
+    //           = p_i * (delta - p_j) where delta is Kronecker delta
+    //
+    // Replacing the p of dFL/dp with p_i, then
+    // dFL/dx_j = dFL/dp_i * dp_i/dx_j
+    //          = - (delta - p_j) * [ (1-p_i)^gamma
+    //              - gamma * (1-p_i)^(gamma-1) * log(p) * p_i]
+    //          =   (delta - p_j) * [- (1-p_i)^gamma +
+    //                gamma * (1-p_i)^(gamma-1) * log(p) * p_i]
+    //
+    // Let B_i denote [- (1-p_i)^gamma +
+    //   gamma * (1-p_i)^(gamma-1) * log(p) * p_i],
+    // and indices {i} is summed for all classes at index j
+    // since x_j received all the gradients from {p_i}.
+    // Then, dFL/dx_j = sum_i{ (delta - p_j) * B_i }
+    //                = sum_i{ (delta*B_i - p_j*B_i }
+    //                = B_j - (p_j * sum_i{B_i})
+
   CUDA_1D_KERNEL_LOOP(index, nthreads) {
-    int64_t label = target[index];
-    T pred = softmax[index * num_classes + label];
+    // B_i
+    const int n = index / num_classes;
+    const int c = index % num_classes;
+
+    const int64_t label = target[n];
+    if (c == label) {
+      const T w = (weight != NULL) ? weight[label] : 1;
+      const T alpha_fac = ((label == 0) * (1 - alpha) + (label >= 1) * alpha) * w;
 
-    if (label >= 0) {
-      buff[index] = alpha * (-pow((T)1. - pred, gamma) +
-                             gamma * pow((T)1. - pred, gamma - 1) * pred *
-                                 log(max(pred, (T)FLT_MIN)));
+      const T log_pred = log_softmax_prob[index];
+      const T pred = exp(log_pred);
+      const T one_minus_pred = 1 - pred;
+
+      grad_input[index] =
+          alpha_fac * (
+              -pow(one_minus_pred, gamma) +
+              gamma * pow(one_minus_pred, gamma - 1) * log_pred * pred
+            );
     } else {
-      buff[index] = 0;
-    }
-    if (weight != NULL) {
-      buff[index] *= weight[label];
+      grad_input[index] = 0;
     }
   }
-}
 
-template <typename T>
-__global__ void softmax_focal_loss_backward_cuda2_kernel(
-    const int nthreads, const T* softmax, const int64_t* target, const T* buff,
-    T* grad_input, const int num_classes) {
   CUDA_1D_KERNEL_LOOP(index, nthreads) {
-    int n = index / num_classes;
-    int c = index % num_classes;
-    int64_t label = target[n];
+    // dFL/dx_j
+    const int n = index / num_classes;
 
-    if (label >= 0) {
-      T flag = (label == c ? (T)1. : (T)0.);
-      grad_input[index] = buff[n] * (flag - softmax[index]);
-    } else {
-      grad_input[index] = 0;
-    }
+    const int start = n * num_classes;
+    const int end = start + num_classes;
+
+    const T sum_buff_along_class = [&] {
+      T sum_buff_along_class = 0;
+      for(int c_idx = start; c_idx < end; ++c_idx) {
+        sum_buff_along_class += grad_input[c_idx];
```

Comment:
I think read grad_input here is not safe.
This thread write `grad_input[index]` and read `grad_input[c_idx]` where c_idx is in range `(n*num_class, n*num_classes+num_classes)`, there is no guarantee that grad_input has been prepared by other threads.